File: source/kl/graphics/Color.java
package kl.graphics;

import kl.math.*;

public class Color extends java.awt.Color {
    public Color() {
        super(0, 0, 0, 255);
    }
    public Color(int rgba) {
        super(rgba, true);
    }
    public Color(int r, int g, int b) {
        super(r, g, b, 255);
    }
    public Color(int r, int g, int b, int a) {
        super(r, g, b, a);
    }
    public Color(float r, float g, float b) {
        super(MathExt.minmax(r, 0.0f, 1.0f),
                MathExt.minmax(g, 0.0f, 1.0f),
                MathExt.minmax(b, 0.0f, 1.0f),
                1.0f);
    }
    public Color(float r, float g, float b, float a) {
        super(MathExt.minmax(r, 0.0f, 1.0f),
                MathExt.minmax(g, 0.0f, 1.0f),
                MathExt.minmax(b, 0.0f, 1.0f),
                MathExt.minmax(a, 0.0f, 1.0f));
    }
    public Color (Int3 values) {
        this(values.x, values.y, values.z);
    }
    public Color (Int4 values) {
        this(values.x, values.y, values.z, values.w);
    }
    public Color (Float3 values) {
        this(values.x, values.y, values.z);
    }
    public Color (Float4 values) {
        this(values.x, values.y, values.z, values.w);
    }

    public boolean equals(Color color) {
        return getRed() == color.getRed() &&
                getBlue() == color.getBlue() &&
                getGreen() == color.getGreen() &&
                getAlpha() == color.getAlpha();
    }

    public Color gray() {
        final int grayValue = (int)(getRed() * 0.3 + getGreen() * 0.585 + getBlue() * 0.115);
        return new Color(grayValue, grayValue, grayValue, getAlpha());
    }

    public Color invert() {
        return new Color(255 - getRed(), 255 - getGreen(), 255 - getBlue(), getAlpha());
    }

    private static final double tableGetter = 9.0 / 255.0;
    private final char[] asciiTable = { '@', '%', '#', 'x', '+', '=', ':', '-', '.', ' ' };
    public char ascii() {
        return asciiTable[(int)(gray().getRed() * tableGetter)];
    }

    private static final double alphaMix = 1.0 / 255.0;
    public Color mix(Color color) {
        return mix(color, color.getAlpha() * alphaMix);
    }
    public Color mix(Color color, double ratio) {
        ratio = Math.min(Math.max(ratio, 0.0), 1.0);
        final double iratio = 1.0 - ratio;
        return new Color(
                (int)(getRed() * iratio) + (int)(color.getRed() * ratio),
                (int)(getGreen() * iratio) + (int)(color.getGreen() *  ratio),
                (int)(getBlue() * iratio) + (int)(color.getBlue() * ratio)
	    );
    }

    public String toString() {
        return "[" + getRed() + ", " + getGreen() + ", " + getBlue() + ", " + getAlpha() + "]";
    }
}

File: Source/Graphics/Color.cpp
#include "graphics/color.h"
#include "math/math.h"


kl::Color::Color() : b(0), g(0), r(0), a(255) {}

kl::Color::Color(byte r, byte g, byte b, byte a) : b(b), g(g), r(r), a(a) {}

bool kl::Color::operator==(const Color& obj) const {
	return r == obj.r && g == obj.g && b == obj.b && a == obj.a;
}

bool kl::Color::operator!=(const Color& obj) const {
	return !(*this == obj);
}

kl::Color kl::Color::asGray() const {
	const byte grayValue = byte(r * 0.3f + g * 0.585f + b * 0.115f);
	return Color(grayValue, grayValue, grayValue, a);
}

kl::Color kl::Color::asInverted() const {
	return Color(255 - r, 255 - g, 255 - b, a);
}

char kl::Color::asASCII() const {
	static const char asciiTable[10] = { '@', '%', '#', 'x', '+', '=', ':', '-', '.', ' ' };
	static const float toSatur = 9.0f / 255.0f;
	return asciiTable[int(asGray().r * toSatur)];
}

kl::Color kl::Color::mix(const Color& col, float ratio) const {
	ratio = Math::MinMax(ratio, 0.0f, 1.0f);
	return Color(
		byte(r * (1.0f - ratio)) + byte(col.r * ratio),
		byte(g * (1.0f - ratio)) + byte(col.g * ratio),
		byte(b * (1.0f - ratio)) + byte(col.b * ratio)
	);
}

kl::Color kl::Color::mix(const Color& col) const {
	return mix(col, ToFloatColor(col.a));
}

std::ostream& kl::operator<<(std::ostream& os, const Color& obj) {
	os << "\033[38;2;" << int(obj.r) << ";" << int(obj.g) << ";" << int(obj.b) << "m";
	return os;
}
